File: java/solutions/src/test/java/org/artrev/workshop/fp/part1/practice3/ListsTest.java
package org.artrev.workshop.fp.part1.practice3;

import org.artrev.workshop.fp.tools.Predicate;
import org.artrev.workshop.fp.tools.predicates.Predicates;
import org.junit.Test;
import org.junit.experimental.runners.Enclosed;
import org.junit.runner.RunWith;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

@RunWith(Enclosed.class)
public class ListsTest {
    public static class Select {
        @Test
        public void should_return_empty_List_for_empty_List() {
            // given:
            final List<String> values = Collections.emptyList();

            // when:
            final List<String> actual =
                    Lists.select(Predicates.alwaysTrue(), values);

            // then:
            assertTrue(actual.isEmpty());
        }

        @Test
        public void should_return_empty_List_if_no_element_satisfies_Predicate() {
            // given:
            final List<String> values = Arrays.asList("JUG", "Lodz", "Workshop");

            // when:
            final List<String> actual =
                    Lists.select(Predicates.alwaysFalse(), values);

            // then:
            assertTrue(actual.isEmpty());
        }

        @Test
        public void should_return_only_elements_that_satisfy_the_Predicate() {
            // given:
            final List<String> values = Arrays.asList("JUG", "Lodz", "Workshop");
            final Predicate<String> predicate = str -> str.length() == 3;
            final String expected = "JUG";

            // when:
            final List<String> actual = Lists.select(predicate, values);

            // then:
            assertEquals(1, actual.size());
            assertEquals(expected, actual.get(0));
        }

        @Test
        public void shouldReturnAllElementsIfAllSatisfyThePredicate() {
            // given:
            final List<String> values = Arrays.asList("JUG", "Lodz", "Workshop");
            final Predicate<String> predicate = Predicates.alwaysTrue();

            // when:
            final List<String> actual = Lists.select(predicate, values);

            // then:
            assertEquals(values, actual);
        }
    }

    public static class Reject {
        @Test
        public void should_return_empty_List_for_empty_List() {
            // given:
            final List<String> values = Collections.emptyList();

            // when:
            final List<String> actual =
                    Lists.select(Predicates.alwaysTrue(), values);

            // then:
            assertTrue(actual.isEmpty());
        }

        @Test
        public void should_return_empty_List_if_ALL_elements_satisfy_the_Predicate() {
            // given:
            final List<String> values = Arrays.asList("JUG", "Lodz", "Workshop");
            final Predicate<String> predicate = str -> str.length() > 2;

            // when:
            final List<String> actual = Lists.reject(predicate, values);

            // then:
            assertTrue(actual.isEmpty());
        }

        @Test
        public void shouldReturnAllElementsThatDoNotSatisfyThePredicate() {
            // given:
            final List<String> values = Arrays.asList("JUG", "Lodz", "Workshop");
            final Predicate<String> predicate = str -> str.length() > 3;
            final String expected = "JUG";

            // when:
            final List<String> actual = Lists.reject(predicate, values);

            // then:
            assertEquals(1, actual.size());
            assertEquals(expected, actual.get(0));
        }

        @Test
        public void shouldReturnAllElementsIfAllDoNotSatisfyThePredicate() {
            // given:
            final List<String> values = Arrays.asList("JUG", "Lodz", "Workshop");

            // when:
            final List<String> actual = Lists.reject(Predicates.alwaysFalse(), values);

            // then:
            assertEquals(values, actual);
        }
    }
}

File: kotlin/solutions/src/test/kotlin/io/github/ajoz/workshop/fp/part1/practice3/ListsTest.kt
package io.github.ajoz.workshop.fp.part1.practice3

import io.github.ajoz.workshop.fp.tools.predicates.alwaysFalse
import io.github.ajoz.workshop.fp.tools.predicates.alwaysTrue
import org.junit.Test
import org.junit.experimental.runners.Enclosed
import org.junit.runner.RunWith

import java.util.Arrays

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue

@RunWith(Enclosed::class)
class ListsTest {
    class Select {
        @Test
        fun `should return empty List for empty List`() {
            // given:
            val values = emptyList<String>()

            // when:
            val actual = Lists.select(alwaysTrue(), values)

            // then:
            assertTrue(actual.isEmpty())
        }

        @Test
        fun `should return empty List if no element satisfies predicate`() {
            // given:
            val values = listOf("JUG", "Lodz", "Workshop")

            // when:
            val actual = Lists.select(alwaysFalse(), values)

            // then:
            assertTrue(actual.isEmpty())
        }

        @Test
        fun `should return only elements that satisfy the predicate`() {
            // given:
            val values = listOf("JUG", "Lodz", "Workshop")
            val predicate: (String) -> Boolean = { it.length == 3 }
            val expected = "JUG"

            // when:
            val actual = Lists.select(predicate, values)

            // then:
            assertEquals(1, actual.size)
            assertEquals(expected, actual[0])
        }

        @Test
        fun `should return ALL elements if ALL satisfy the predicate`() {
            // given:
            val values = listOf("JUG", "Lodz", "Workshop")

            // when:
            val actual = Lists.select(alwaysTrue(), values)

            // then:
            assertEquals(values, actual)
        }
    }

    class Reject {
        @Test
        fun `should return empty List for empty List`() {
            // given:
            val values = emptyList<String>()

            // when:
            val actual = Lists.select(alwaysTrue(), values)

            // then:
            assertTrue(actual.isEmpty())
        }

        @Test
        fun `should return empty List if ALL elements satisfy the predicate`() {
            // given:
            val values = listOf("JUG", "Lodz", "Workshop")
            val predicate: (String) -> Boolean = { it.length > 2 }

            // when:
            val actual = Lists.reject(predicate, values)

            // then:
            assertTrue(actual.isEmpty())
        }

        @Test
        fun `should return ALL elements that do not satisfy the predicate`() {
            // given:
            val values = Arrays.asList("JUG", "Lodz", "Workshop")
            val predicate: (String) -> Boolean = { it.length > 3 }
            val expected = "JUG"

            // when:
            val actual = Lists.reject(predicate, values)

            // then:
            assertEquals(1, actual.size.toLong())
            assertEquals(expected, actual[0])
        }

        @Test
        fun `should return ALL elements if ALL do not satisfy the predicate`() {
            // given:
            val values = Arrays.asList("JUG", "Lodz", "Workshop")

            // when:
            val actual = Lists.reject(alwaysFalse(), values)

            // then:
            assertEquals(values, actual)
        }
    }
}
